### ISSUE TITLE
C# Tool-/EditorScript can only be run successfully once from script editor (ctrl+shift+x).

### ISSUE THREAD

**Issue by dazKind** [godotengine/godot#48590]
**Godot version:**
```
Godot Engine v3.3.1.rc.mono.custom_build.476bc5191
commit 476bc5191bafcdea99e918100394479c06f0e5b4 (HEAD -> 3.3, origin/3.3)
Author: Tomasz Chabora <kobewi4e@gmail.com>
Date:   Fri Mar 26 14:10:41 2021 +0100
    Save project after opening
...
```
also tested on
```
Godot Engine v3.3.stable.mono.official - https://godotengine.org
```

**OS/device including version:**
```
Windows 10
Visual Studio 2017 - Found MSVC compiler: amd64
Microsoft (R) Build Engine version 16.9.0+57a23d249 for .NET
Mono JIT compiler version 6.12.0 (Visual Studio built mono)
```

**Issue description**
I'm currently evaluating Godot+Mono for porting over a project that comes with a lot of C# code and I'm taking a closer look at procedural content creation in the editor through c#-based tool-scripts. Unfortunately I hit a wall when it comes to reliably running C# based tool-scripts. I can run them successfully once and then i need to restart the editor to make them work again. Comparing this to GDScript based tool-scripts, this seems clearly broken.

Here are the specifics based on a minimal repro-sample I include at the end of this post:

I have 2 tool editorscripts, one GDscript and one C# script.
I can run the GDScript from the script-editor (ctrl+shift+x) without problems multiple times and everything works as expected.
I can run the C# script from the script-editor (ctrl+shift+x) exactly once. Starting with the second run the editor prints:
```
ERROR: CSharpScript::reload: Condition "!p_keep_state && has_instances" is true. Returned: ERR_ALREADY_IN_USE
   At: modules\mono\csharp_script.cpp:3160
```

Upon further investigation & after reviewing the contents of the instances container [here](https://github.com/godotengine/godot/blob/3.3/modules/mono/csharp_script.cpp#L3157), I found that there is an EditorScript owner left over which prevents the script from reloading and throwing the above mentioned error. 

My suspicious is that this could be due to an edge-case or even a leak concerning c# editor-scripts that fell through the cracks so far. Possibly related to the script-editor itself. 

Before I start hacking away at this to see if this could be some low-hanging fruit, it still requires a lot of domain knowledge concerning the C# module that I don't have right now. So, I'd love to have your thoughts and input, maybe together we can easily zero down on this further.

**Steps to reproduce:**
1. Open the supplied project "Script Debug"
2. On the lower left in the resource-browser, double click `tool_test.gd` to open it in the script-editor
3. (ctrl+shift+x) to run the script at least 2 times and see the prints in the editor console:
```
Hello from tool_test.gd
Hello from tool_test.gd
```
4. On the lower left in the resource-browser, double click `tool_test.cs` to open it in the script-editor
5. Run a C# build: "MSBuild->Build Solution"
6. In the script-editor (ctrl+shift+x) to run the c# script at least 2 times and see the prints in the editor console:
```
Hello from tool_test.cs
ERROR: CSharpScript::reload: Condition "!p_keep_state && has_instances" is true. Returned: ERR_ALREADY_IN_USE
   At: modules\mono\csharp_script.cpp:3160
```

**Minimal reproduction project:**
[ScriptDebug.zip](https://github.com/godotengine/godot/files/6448002/ScriptDebug.zip)

**Comment by dazKind** [comment]
Ok, I think the issue is related to this [EditorScript Reference](https://github.com/godotengine/godot/blob/master/editor/plugins/script_editor_plugin.cpp#L1321) and how it interacts with Mono's Garbage Collection. 

I just managed to catch a GC-cycle after where the instance got cleared after I ran it the first time. Upon completing the cycle I could see the instances were empty and I was able to reload & run the script a second time in one editor session.

Now I setup a finer net of breakpoints and I'm trying to catch another cycle to gain more insights. Will post the results here as soon as I got them.

**Comment by dazKind** [comment]
Ok, caught it again. Indeed, it takes ages for mono to clear the EditorScript reference! After this happens, the script can be run again.
```
 	godot.windows.tools.64.mono.exe!CSharpInstance::~CSharpInstance() Line 2199	C++	Symbols loaded.
 	godot.windows.tools.64.mono.exe!CSharpInstance::`scalar deleting destructor'()	C++	Non-user code. Symbols loaded.
 	godot.windows.tools.64.mono.exe!memdelete<ScriptInstance>(p_class=0x0000025e28fed790) Line 119	C++	Symbols loaded.
 	godot.windows.tools.64.mono.exe!Object::~Object() Line 2007	C++	Symbols loaded.
 	godot.windows.tools.64.mono.exe!Reference::~Reference() Line 112	C++	Symbols loaded.
 	godot.windows.tools.64.mono.exe!EditorScript::~EditorScript()	C++	Non-user code. Symbols loaded.
 	godot.windows.tools.64.mono.exe!EditorScript::`scalar deleting destructor'()	C++	Non-user code. Symbols loaded.
 	godot.windows.tools.64.mono.exe!memdelete<Reference>(p_class=0x0000025e3a93f5c0) Line 119	C++	Symbols loaded.
>	godot.windows.tools.64.mono.exe!godot_icall_Reference_Disposed(p_obj=0x0000025e29f6cea0, p_ptr=0x0000025e3a93f5c0, p_is_finalizer='\x1') Line 100	C++	Symbols loaded.
```

Now on to finding out what can be done about this.

**Comment by Calinou** [comment]
Can the EditorScript instance be manually disposed? See https://github.com/godotengine/godot/issues/41543 which is a similar issue for InputEvent.

**Comment by dazKind** [comment]
Any pointers on how would I go about that?

**Comment by Calinou** [comment]
> Any pointers on how would I go about that?

This is likely something that would have to be done from the editor C# code itself.

**Comment by dazKind** [comment]
Ok, I took a closer look to understand how the reference counting works in relation to the C# garbage collection.
It looks like a C# Instance holds on to the cpp reference till a GC cycle comes along upon which the instance and the reference will get cleared for good. 

Now, this means script instances which have been created and released from the cpp code will till retain a reference with a count of 1. Further, if we are naive and brave, we could use this as a identification condition and adapt the reload logic of C# scripts to only complain about instances that have a refcount > 1. 
Ofc, this means we would collect objects till the next GC cycle, but I think this is perfectly fine for the use-case of tool scripts.

I added this to my fork and I'm testing this right now and it looks like it works just fine. Will test this with a bigger project now.
